File: Reporting Services/Extension Samples/PolygonCustomReportItem Sample/CS/PolygonsDesigner/PolygonProperties.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//  This source code is intended only as a supplement to Microsoft
//  Development Tools and/or on-line documentation.  See these other
//  materials for detailed information regarding Microsoft code samples.
// 
//  THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//  PARTICULAR PURPOSE.
//-----------------------------------------------------------------------

namespace Microsoft.Samples.ReportingServices
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.Data;
    using System.Drawing;
    using System.Text;
    using System.Windows.Forms;
    using Microsoft.ReportDesigner;
    using Microsoft.ReportDesigner.Design;
    using Microsoft.ReportingServices.Interfaces;
    using Rdl = Microsoft.ReportingServices.RdlObjectModel;

    public partial class PolygonProperties : Form
    {
        private PolygonsDesigner designerComponent;
        private Rdl.DataSet reportDataSet;
        private String oldComboValue;
        private bool launchEditor;

        public PolygonProperties()
        {
            InitializeComponent();
        }

        [System.CLSCompliant(false)]
        public PolygonsDesigner DesignerComponent
        {
            get
            {
                return this.designerComponent;
            }

            set
            {
                this.designerComponent = value;
            }
        }

        private void PolygonProperties_Load(object sender, EventArgs e)
        {
            this.DataSetName.Text = this.designerComponent.DataSetName;
            IList<Rdl.DataSet> datasets = this.designerComponent.Report.DataSets;
            for (int i = 0; i < datasets.Count; i++)
            {
                this.DataSetName.Items.Add(datasets[i].Name);
            }
            
            this.UpdateDataSet();
            this.ShapeColor.Text = this.designerComponent.ShapeColor;
            ProportionalScaling.Checked = (
                this.designerComponent.GetCustomProperty("poly:Proportional") == bool.TrueString);
            Translucency.Text = this.designerComponent.Translucency;
            this.MinX.Text = this.designerComponent.GetCustomProperty("poly:MinX");
            this.MaxX.Text = this.designerComponent.GetCustomProperty("poly:MaxX");
            this.MinY.Text = this.designerComponent.GetCustomProperty("poly:MinY");
            this.MaxY.Text = this.designerComponent.GetCustomProperty("poly:MaxY");
            this.PointID.Text = this.designerComponent.PointExpression;
            this.ShapeID.Text = this.designerComponent.ShapeExpression;
            this.XValue.Text = this.designerComponent.XExpression;
            this.YValue.Text = this.designerComponent.YExpression;
            this.Hyperlink.Text = this.designerComponent.ShapeHyperlink;
        }

        private void OK_Click(object sender, EventArgs e)
        {
            this.designerComponent.DataSetName = this.DataSetName.Text;
            this.designerComponent.ShapeColor = this.ShapeColor.Text;
            this.designerComponent.SetCustomProperty(
                "poly:Proportional", 
                ProportionalScaling.Checked ? bool.TrueString : bool.FalseString);
            this.designerComponent.Translucency = Translucency.Text;
            this.designerComponent.SetCustomProperty("poly:MinX", MinX.Text);
            this.designerComponent.SetCustomProperty("poly:MaxX", MaxX.Text);
            this.designerComponent.SetCustomProperty("poly:MinY", MinY.Text);
            this.designerComponent.SetCustomProperty("poly:MaxY", MaxY.Text);
            this.designerComponent.PointExpression = PointID.Text;
            this.designerComponent.ShapeExpression = ShapeID.Text;
            this.designerComponent.XExpression = XValue.Text;
            this.designerComponent.YExpression = YValue.Text;
            this.designerComponent.ShapeHyperlink = this.Hyperlink.Text;
        }

        private void DataSetName_SelectedIndexChanged(object sender, EventArgs e)
        {
            this.UpdateDataSet();
        }

        private void UpdateDataSet()
        {
            this.reportDataSet = null;
            IList<Rdl.DataSet> datasets = this.designerComponent.Report.DataSets;
            for (int i = 0; i < datasets.Count; i++)
            {
                if (datasets[i].Name == this.DataSetName.Text)
                {
                    this.reportDataSet = datasets[i];
                    break;
                }
            }

            this.PopulateFieldSelector(this.ShapeID);
            this.PopulateFieldSelector(this.ShapeColor);
            this.PopulateFieldSelector(this.Hyperlink);
            this.PopulateFieldSelector(this.PointID);
            this.PopulateFieldSelector(this.XValue);
            this.PopulateFieldSelector(this.YValue);
        }

        private void PopulateFieldSelector(ComboBox combo)
        {
            combo.Items.Clear();
            combo.Items.Add("<Expression...>");
            if (this.reportDataSet == null)
            {
                return;
            }

            for (int i = 0; i < this.reportDataSet.Fields.Count; i++)
            {
                combo.Items.Add("=Fields!" + this.reportDataSet.Fields[i].Name + ".Value");
            }
        }

        private void EditableCombo_DropDown(object sender, EventArgs e)
        {
            ComboBox combo = (ComboBox)sender;
            if (combo.SelectedIndex == 0)
            {
                string currentValue = combo.Text;
                combo.SelectedIndex = -1;
                combo.Items[0] = "<Expression...>";
                combo.Text = currentValue;
            }
            else
            {
                combo.Items[0] = "<Expression...>";
            }
        }

        private void EditableCombo_SelectionChangeCommitted(object sender, EventArgs e)
        {
            ComboBox combo = (ComboBox)sender;
            this.oldComboValue = combo.Text;
            this.launchEditor = true;
        }

        private void EditableCombo_SelectedIndexChanged(object sender, EventArgs e)
        {
            ComboBox combo = (ComboBox)sender;
            if (combo.SelectedIndex == 0 && this.launchEditor)
            {
                this.launchEditor = false;

                // if <Expression...> is selected in the combo box,
                // invoke the report builder expression editor
                ExpressionEditor editor = new ExpressionEditor();
                string newValue;
                newValue = (string)editor.EditValue(null, this.designerComponent.Site, new Rdl.ReportExpression(this.oldComboValue));
                combo.Items[0] = newValue;
            }
        }
    }
}

File: Reporting Services/Extension Samples/PolygonCustomReportItem Sample/VB/PolygonsDesigner/PolygonProperties.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
' 
'  THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'-----------------------------------------------------------------------
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports Microsoft.ReportDesigner
Imports Microsoft.ReportDesigner.Design
Imports Microsoft.ReportingServices.Interfaces
Imports Rdl = Microsoft.ReportingServices.RdlObjectModel


Partial Public Class PolygonProperties
    Inherits Form
    Private component As PolygonsDesigner
    Private reportDataSet As Rdl.DataSet
    Private oldComboValue As String
    Private launchEditor As Boolean

    Public Sub New()
        InitializeComponent()
    End Sub

    <System.CLSCompliant(False)> _
    Public Property DesignerComponent() As PolygonsDesigner
        Get
            Return Me.component
        End Get

        Set(ByVal value As PolygonsDesigner)
            Me.component = value
        End Set
    End Property

    Private Sub PolygonProperties_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        Me.DataSetName.Text = Me.component.DataSetName
        Dim datasets As IList(Of Rdl.DataSet) = New List(Of Rdl.DataSet)
        For i As Integer = 0 To datasets.Count - 1
            Me.DataSetName.Items.Add(datasets(i).Name)
        Next

        Me.UpdateDataSet()
        Me.ShapeColor.Text = Me.component.ShapeColor
        If Me.component.GetCustomProperty("poly:Proportional") = Boolean.TrueString Then
            ProportionalScaling.Checked = True
        Else
            ProportionalScaling.Checked = False
        End If

        Translucency.Text = Me.component.Translucency
        Me.MinX.Text = Me.component.GetCustomProperty("poly:MinX")
        Me.MaxX.Text = Me.component.GetCustomProperty("poly:MaxX")
        Me.MinY.Text = Me.component.GetCustomProperty("poly:MinY")
        Me.MaxY.Text = Me.component.GetCustomProperty("poly:MaxY")
        Me.PointID.Text = Me.component.PointExpression
        Me.ShapeID.Text = Me.component.ShapeExpression
        Me.XValue.Text = Me.component.XExpression
        Me.YValue.Text = Me.component.YExpression
        Me.Hyperlink.Text = Me.component.ShapeHyperlink
    End Sub

    Private Sub OK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles OK.Click
        Me.component.DataSetName = Me.DataSetName.Text
        Me.component.ShapeColor = Me.ShapeColor.Text
        If ProportionalScaling.Checked = True Then
            Me.component.SetCustomProperty("poly:Proportional", Boolean.TrueString)
        Else
            Me.component.SetCustomProperty("poly:Proportional", Boolean.FalseString)
        End If

        Me.component.Translucency = Translucency.Text
        Me.component.SetCustomProperty("poly:MinX", MinX.Text)
        Me.component.SetCustomProperty("poly:MaxX", MaxX.Text)
        Me.component.SetCustomProperty("poly:MinY", MinY.Text)
        Me.component.SetCustomProperty("poly:MaxY", MaxY.Text)
        Me.component.PointExpression = PointID.Text
        Me.component.ShapeExpression = ShapeID.Text
        Me.component.XExpression = XValue.Text
        Me.component.YExpression = YValue.Text
        Me.component.ShapeHyperlink = Me.Hyperlink.Text
    End Sub

    Private Sub DataSetName_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles DataSetName.SelectedIndexChanged
        Me.UpdateDataSet()
    End Sub

    Private Sub UpdateDataSet()
        Me.reportDataSet = Nothing
        Dim datasets As IList(Of Rdl.DataSet) = New List(Of Rdl.DataSet)
        For i As Integer = 0 To datasets.Count - 1
            If datasets(i).Name = Me.DataSetName.Text Then
                Me.reportDataSet = datasets(i)
                Exit For
            End If
        Next

        Me.PopulateFieldSelector(Me.ShapeID)
        Me.PopulateFieldSelector(Me.ShapeColor)
        Me.PopulateFieldSelector(Me.Hyperlink)
        Me.PopulateFieldSelector(Me.PointID)
        Me.PopulateFieldSelector(Me.XValue)
        Me.PopulateFieldSelector(Me.YValue)
    End Sub

    Private Sub PopulateFieldSelector(ByVal combo As ComboBox)
        combo.Items.Clear()
        combo.Items.Add("<Expression...>")
        If Me.reportDataSet Is Nothing Then
            Return
        End If

        For i As Integer = 0 To (Me.reportDataSet.Fields.Count) - 1
            combo.Items.Add(("=Fields!" & Me.reportDataSet.Fields(i).Name & ".Value"))
        Next
    End Sub

    Private Sub EditableCombo_DropDown(ByVal sender As Object, ByVal e As EventArgs) Handles Hyperlink.DropDown, PointID.DropDown, ShapeID.DropDown, XValue.DropDown, YValue.DropDown, Translucency.DropDown, ShapeColor.DropDown
        Dim combo As ComboBox = CType(sender, ComboBox)
        If combo.SelectedIndex = 0 Then
            Dim currentValue As String = combo.Text
            combo.SelectedIndex = -1
            combo.Items(0) = "<Expression...>"
            combo.Text = currentValue
        Else
            combo.Items(0) = "<Expression...>"
        End If
    End Sub

    Private Sub EditableCombo_SelectionChangeCommitted(ByVal sender As Object, ByVal e As EventArgs) Handles Hyperlink.SelectionChangeCommitted, PointID.SelectionChangeCommitted, ShapeID.SelectionChangeCommitted, XValue.SelectionChangeCommitted, YValue.SelectionChangeCommitted, Translucency.SelectionChangeCommitted, ShapeColor.SelectionChangeCommitted
        Dim combo As ComboBox = CType(sender, ComboBox)
        Me.oldComboValue = combo.Text
        Me.launchEditor = True
    End Sub

    Private Sub EditableCombo_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles Hyperlink.SelectedIndexChanged, PointID.SelectedIndexChanged, ShapeID.SelectedIndexChanged, XValue.SelectedIndexChanged, YValue.SelectedIndexChanged, Translucency.SelectedIndexChanged, ShapeColor.SelectedIndexChanged
        Dim combo As ComboBox = CType(sender, ComboBox)
        If combo.SelectedIndex = 0 And Me.launchEditor Then
            Me.launchEditor = False

            ' if <Expression...> is selected in the combo box,
            ' invoke the report builder expression editor
            Dim editor As New ExpressionEditor()
            Dim newValue As String
            newValue = CStr(editor.EditValue(Nothing, Me.component.Site, Me.oldComboValue))
            combo.Items(0) = newValue
        End If
    End Sub
End Class
